In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import glob 
import os
import warnings
import networkx as nx
import matplotlib.colors as mcolors
warnings.simplefilter(action='ignore', category=FutureWarning)
from scipy.stats import pearsonr
from sklearn.covariance import GraphicalLassoCV
from networkx.generators.community import gaussian_random_partition_graph
import scipy.stats as stats
from scipy.stats import ttest_ind
import nibabel as nib 
import seaborn as sns
from sklearn.preprocessing import StandardScaler
from tqdm import tqdm
pd.options.display.max_rows = None
pd.options.display.max_columns = None
import glob as glob
import math
#set random generator outside loop 
#seed everytrhing once at begibning and dont touch it 
np.random.seed(3)

#### For tau graphs, just do these steps, but on Tau suvr, stratify by amyloid quartile 

- For stratifying by amyloid status,  Royse 2021 Validation of PET positivity thresholds in centiloids established that 21 (or 20) is about so use this

#### Data has been preprocessed into amyloid quartiles for both adni and A4

## I need to verify they used the same freesurfer atlases (for region specific analyses)


#### General steps 
#### 1. A. Read ADNI data quartile csvs 
- These data have regional tau SUVRs, centiloid values, visit dates, and ID 
- (103, 228) in low 
- (107, 228) in med low 
- (100, 228) in med high 
- (101, 228) in high 

In [10]:
adni_cent_low = pd.read_csv('./adni/adni_amy_tau_merged_cent_low_quartile.csv')
adni_cent_med_low = pd.read_csv('./adni/adni_amy_tau_merged_cent_med_low_quartile.csv')
adni_cent_med_high = pd.read_csv('./adni/adni_amy_tau_merged_cent_med_high_quartile.csv')
adni_cent_high = pd.read_csv('./adni/adni_amy_tau_merged_cent_high_quartile.csv')

In [14]:
adni_cent_low.head(5)

RID    SCANDATE TRACER  CENTILOIDS VISCODE VISCODE2    EXAMDATE  \
0   120  2017-09-26    FBP        32.0    init     m132  2017-09-13   
1   677  2019-08-08    FBP        45.0      y2     m156  2019-09-17   
2   896  2018-03-14    FBP        35.0    init     m138  2018-11-14   
3  1280  2017-03-02    FBP        33.0      y1     m132  2018-03-05   
4  2245  2021-01-21    FBP        31.0      y4     m120  2021-01-20   

   META_TEMPORAL_SUVR  META_TEMPORAL_VOLUME  CTX_ENTORHINAL_SUVR  \
0               1.671               68181.0                1.894   
1               1.839               62887.0                1.834   
2               1.800               57175.0                2.969   
3               1.716               54287.0                2.350   
4               1.642               61602.0                2.114   

   CTX_ENTORHINAL_VOLUME  INFERIOR_CEREBGM_SUVR  INFERIOR_CEREBGM_VOLUME  \
0                 3441.0                    1.0                  69868.0   
1                 3567.0                    1.0                  64186.0   
2                 2386.0                    1.0                  73645.0   
3                 3025.0                    1.0                  58083.0   
4                 3271.0                    1.0                  65177.0   

   HEMIWM_SUVR  HEMIWM_VOLUME  BRAIN_STEM_SUVR  BRAIN_STEM_VOLUME  \
0        0.833       436301.0            0.973            22723.0   
1        1.171       397130.0            1.063            20832.0   
2        0.948       402452.0            0.993            21978.0   
3        0.788       354571.0            1.016            19588.0   
4        0.829       364197.0            1.130            19377.0   

   CHOROID_SUVR  CHOROID_VOLUME  MIDDLEFR_SUVR  MIDDLEFR_VOLUME  \
0         5.876          1424.0          1.704          39203.0   
1         4.081          2033.0          1.971          40278.0   
2         4.577          1943.0          1.697          48449.0   
3         3.831          1835.0          1.818          28326.0   
4         3.434          1343.0          1.553          33343.0   

   ORBITOFR_SUVR  ORBITOFR_VOLUME  PARSFR_SUVR  PARSFR_VOLUME  \
0          1.718          25745.0        1.828        18739.0   
1          1.881          27123.0        2.138        21217.0   
2          1.839          31600.0        1.694        20560.0   
3          1.708          22150.0        1.823        17309.0   
4          1.587          25240.0        1.524        21455.0   

   CTX_BANKSSTS_SUVR  CTX_BANKSSTS_VOLUME  CTX_CAUDALANTERIORCINGULATE_SUVR  \
0              1.539               4407.0                             1.611   
1              1.586               3333.0                             1.681   
2              1.471               4579.0                             1.326   
3              1.473               3300.0                             1.222   
4              1.451               3406.0                             1.280   

   CTX_CAUDALANTERIORCINGULATE_VOLUME  CTX_CUNEUS_SUVR  CTX_CUNEUS_VOLUME  \
0                              3260.0            1.773             6121.0   
1                              3233.0            1.806             5063.0   
2                              4624.0            1.690             6995.0   
3                              3161.0            1.694             5398.0   
4                              2884.0            1.389             5939.0   

   CTX_FUSIFORM_SUVR  CTX_FUSIFORM_VOLUME  CTX_INFERIORPARIETAL_SUVR  \
0              1.494              18478.0                      1.820   
1              1.655              19354.0                      2.159   
2              1.526              16040.0                      1.737   
3              1.433              14790.0                      1.929   
4              1.433              16489.0                      1.602   

   CTX_INFERIORPARIETAL_VOLUME  CTX_INFERIORTEMPORAL_SUVR  \
0                      26934.0                      1.733   
1        

#### 1. B. Read A4 data 
- These data have regional tau SUVRs, centiloid values, visit dates, and ID 
- (101, 277) in low 
- (86, 277) in med low 
- (94, 277) in med high
- (92, 277) in high

In [15]:
a4_cent_low = pd.read_csv('./a4/a4_amy_tau_merged_cent_low_quartile.csv')
a4_cent_med_low = pd.read_csv('./a4/a4_amy_tau_merged_cent_med_low_quartile.csv')
a4_cent_med_high = pd.read_csv('./a4/a4_amy_tau_merged_cent_med_high_quartile.csv')
a4_cent_high = pd.read_csv('./a4/a4_adni_amy_tau_merged_cent_high_quartile.csv')

In [17]:
a4_cent_low.head(5)

BID         update_stamp_x       ligand  centiloid  \
0  B10709603  2018-12-06 00:00:00.0  Florbetapir    40.5933   
1  B11089276  2018-12-06 00:00:00.0  Florbetapir    35.1012   
2  B12755986  2018-12-06 00:00:00.0  Florbetapir    33.2705   
3  B14928511  2018-12-06 00:00:00.0  Florbetapir    22.2863   
4  B16052018  2018-12-06 00:00:00.0  Florbetapir    36.9319   

   Mean_3rd_Ventricle  Mean_4th_Ventricle  Mean_Brain_Stem  Mean_CC_Anterior  \
0            1.041388            0.869095         1.007361          0.853996   
1            0.960801            1.035359         1.035534          0.875824   
2            0.874286            1.001917         0.914272          0.868799   
3            1.130561            0.864720         1.035799          0.984863   
4            0.894230            0.867076         0.981340          0.983319   

   Mean_CC_Central  Mean_CC_Mid_Anterior  Mean_CC_Mid_Posterior  \
0         0.872356              0.826985               0.782240   
1         0.751278              0.736648               0.804294   
2         0.782858              0.824874               0.746615   
3         0.890723              0.929738               0.792920   
4         0.885003              0.892201               0.899059   

   Mean_CC_Posterior  Mean_CSF  Mean_Left_Accumbens_area  Mean_Left_Amygdala  \
0           0.966260  1.205540                  1.433114            1.317141   
1           0.880089  0.941958                  1.459700            1.271211   
2           0.826709  0.947344                  1.501147            1.111348   
3           1.038879  1.148673                  1.474145            1.316118   
4           1.005557  1.024845                  1.279253            1.271878   

   Mean_Left_Caudate  Mean_Left_Cerebellum_Cortex  \
0           1.153687                     0.998110   
1           1.275619                     1.010331   
2           1.119478                     0.996641   
3           1.131254                     0.987746   
4           1.149294                     0.976525   

   Mean_Left_Cerebellum_White_Matter  Mean_Left_Cerebral_White_Matter  \
0                           1.160842                         1.183678   
1                           1.108437                         1.204417   
2                           1.046534                         1.121524   
3                           1.152828                         1.138394   
4                           1.056285                         1.191551   

   Mean_Left_Hippocampus  Mean_Left_Inf_Lat_Vent  Mean_Left_Lateral_Ventricle  \
0               1.640952                1.702160                     0.820657   
1               1.257985                1.230950                     0.682581   
2               1.207440                1.402119                     0.526838   
3               1.281461                1.256047                     0.695696   
4               1.214709                1.311286                     0.770285   

   Mean_Left_Pallidum  Mean_Left_Putamen  Mean_Left_Thalamus  \
0            1.736637           1.587242            1.369225   
1            1.736647           1.684787            1.397907   
2            1.836690           1.554073            1.300996   
3            1.659153           1.570603            1.298524   
4            1.616788           1.509911            1.194891   

   Mean_Left_VentralDC  Mean_Left_choroid_plexus  Mean_Left_vessel  \
0             1.414810                  2.476843               NaN   
1             1.349388                  1.284944          1.450377   
2             1.353409                  1.364310          1.270827   
3             1.399427                  0.851118          1.549292   
4             1.180558                  1.078686          1.479832   

   Mean_Optic_Chiasm  Mean_Right_Accumbens_area  Mean_Right_Amygdala  \
0           0.746455                   1.340585             1.227425   
1           0.756333                   1.384146             1.208867  

In [2]:
df = pd.read_csv("/home/mlauber/mci_mri_graph/construct_ggm/ml_dataframe.csv")

In [3]:
#look for and load in the A4 dataset because I think it has more patients

In [4]:
print(len(df))
print(df.columns)

793
Index(['RID', 'PTGENDER', 'PTEDUCAT', 'PTTLANG', 'PTETHCAT', 'PTRACCAT',
       'AGE_MRI', 'DIAGNOSIS', 'Q1SCORE', 'Q4SCORE',
       ...
       'MMSE3', 'MMSE4', 'MMSE5', 'MMSE6', 'MMSE7', 'META_TEMPORAL',
       'MESIAL_TEMPORAL', 'TEMPO_PARIETAL', 'FRONTAL', 'NEOCORTICAL'],
      dtype='object', length=202)


- out of 793 patients 
- if cut off is 24 then 
- 339 amy pos
- 454 amy neg

- if cut off is 40 then 
- 256 pos : if I do quartiles then this is 64 patients in each quartile, could be kinda small, maybe use this as the replication dataset if possible 
- 526 neg 

In [11]:
#separate into pos and neg 
amy_pos = df[df['CENTILOIDS']>= 18]
amy_neg = df[df['CENTILOIDS']< 18]

# #if I use a stricter cut off of 40
# amy_pos = df[df['CENTILOIDS']>= 40]
# amy_neg = df[df['CENTILOIDS']< 40]
print(len(amy_pos))
print(len(amy_neg))

371
422


In [12]:
371 / 4

92.75

#### 2. Run and train separate graphical models for amyloid+ and amyloid- populations

##### I. Prepare data steps:
- a. need drop all non brain region vol cols - keep only brain regions 
- also drop the left and right ones 

In [7]:
#do some reg ex search for _VOLUME
def filter_regions(df):
    filtered_df = df.filter(regex='_VOLUME$')
    return filtered_df

amy_pos_reg = filter_regions(amy_pos)
amy_neg_reg = filter_regions(amy_neg)
amy_pos_reg.to_csv('amy_pos_regions.csv')
amy_neg_reg.to_csv('amy_neg_reg.csv')
print(amy_neg_reg.columns)

Index(['META_TEMPORAL_VOLUME', 'CTX_ENTORHINAL_VOLUME',
       'INFERIOR_CEREBGM_VOLUME', 'HEMIWM_VOLUME', 'BRAIN_STEM_VOLUME',
       'CHOROID_VOLUME', 'MIDDLEFR_VOLUME', 'ORBITOFR_VOLUME', 'PARSFR_VOLUME',
       'CTX_BANKSSTS_VOLUME',
       ...
       'LEFT_CAUDATE_VOLUME', 'LEFT_PALLIDUM_VOLUME', 'LEFT_PUTAMEN_VOLUME',
       'LEFT_THALAMUS_PROPER_VOLUME', 'RIGHT_ACCUMBENS_AREA_VOLUME',
       'RIGHT_AMYGDALA_VOLUME', 'RIGHT_CAUDATE_VOLUME',
       'RIGHT_PALLIDUM_VOLUME', 'RIGHT_PUTAMEN_VOLUME',
       'RIGHT_THALAMUS_PROPER_VOLUME'],
      dtype='object', length=110)


In [8]:
print(len(amy_pos.columns))

202


In [9]:
#drop left and right because some are redudant 
#something like if LEFT_ RIGHT_ or LH or RH
def filter_left_right(df):
    columns_to_keep = [col for col in df.columns if 'LEFT' not in col and 'RIGHT' not in col and 'LH' not in col and 'RH' not in col]
    return df[columns_to_keep]

# Assuming you have data in amy_pos_reg and amy_neg_reg
amy_pos_reg_comb = filter_left_right(amy_pos_reg)
amy_neg_reg_comb = filter_left_right(amy_neg_reg)

amy_pos_reg_comb.to_csv('amy_pos_reg_dropleftright.csv')
amy_neg_reg_comb.to_csv('amy_neg_reg_dropleftright.csv')

print(amy_pos_reg_comb.columns)
print(len(amy_pos_reg_comb.columns))



Index(['META_TEMPORAL_VOLUME', 'INFERIOR_CEREBGM_VOLUME', 'HEMIWM_VOLUME',
       'BRAIN_STEM_VOLUME', 'CHOROID_VOLUME', 'MIDDLEFR_VOLUME',
       'ORBITOFR_VOLUME', 'PARSFR_VOLUME', 'CTX_BANKSSTS_VOLUME',
       'CTX_CAUDALANTERIORCINGULATE_VOLUME', 'CTX_CUNEUS_VOLUME',
       'CTX_FUSIFORM_VOLUME', 'CTX_INFERIORPARIETAL_VOLUME',
       'CTX_INFERIORTEMPORAL_VOLUME', 'CTX_INSULA_VOLUME',
       'CTX_ISTHMUSCINGULATE_VOLUME', 'CTX_LATERALOCCIPITAL_VOLUME',
       'CTX_LINGUAL_VOLUME', 'CTX_MIDDLETEMPORAL_VOLUME',
       'CTX_PARACENTRAL_VOLUME', 'CTX_PARAHIPPOCAMPAL_VOLUME',
       'CTX_PERICALCARINE_VOLUME', 'CTX_POSTCENTRAL_VOLUME',
       'CTX_POSTERIORCINGULATE_VOLUME', 'CTX_PRECENTRAL_VOLUME',
       'CTX_PRECUNEUS_VOLUME', 'CTX_ROSTRALANTERIORCINGULATE_VOLUME',
       'CTX_SUPERIORFRONTAL_VOLUME', 'CTX_SUPERIORPARIETAL_VOLUME',
       'CTX_SUPERIORTEMPORAL_VOLUME', 'CTX_SUPRAMARGINAL_VOLUME',
       'CTX_TEMPORALPOLE_VOLUME', 'CTX_TRANSVERSETEMPORAL_VOLUME',
       'ACCUMBENS_ARE

- b. find out if there are any regions which have 0 volume or rows where there is no entry 

- c. normalize the data 

- d. check that there are no null values 

##### II. Create bootstrap samples and run model for amy pos

- a. create boot strap samples 

- b. run model 

- c. calc metrics 

##### III. Create bootstrap samples and run model for amy neg

- a. create boot strap samples 

- b. run model 

- c. calc metrics 

##### IV. Compare graph metrics using t test 

##### Later can Use GMM (2 components) to stratify patients into amyloid pos and neg 
- example code in mri_pet/pet_predicition_mldataframe.csv
- do thresholding on normal cognition 